### PR TITLE
[WIP] [CompilerPerf] trial generating unverifiable code

### PR DESCRIPTION
**Pull request by dsyme**
In this PR, we experimentally add a switch to the F# compiler to generate unverifiable code for some pattern matching.  In particular for some discriminated unions, the `castdata` operation will never, logically speaking, fail.  This means it is valid IL (but unverifiable) to omit the `castclass` operation for those discriminated unions that compile to a base class with subclasses.

This experimentally adds an `--unverifiable` switch and uses it for `FSharp.Compiler.dll`.  Let's see if it improves the perf of the compiler at all.

### CODE REVIEW

**Comment by dsyme** [comment]
CompilerPerfTest please  (nb. this marker is used by the perf testing scripts in this repo)

**Comment by dsyme** [comment]
OK, 

* I've adjusted the compiler perf testing scripts to automatically perf test any PR that has ``[CompilerPerf]`` in the title.  However I still have to manually start the compiler perf testing on the stable configuration machine

* This particular PR does seem to give a marginal performance 3-4% improvement in the type checking and code generation phases of the compiler.  You have to look carefully at [this raw results file](https://github.com/Microsoft/visualfsharp/blob/master/tests/scripts/compiler-perf-results.txt) to see that.  

* A visualization of the data in that file will be very important over time.

**Comment by KevinRansom** [comment]
@dsyme 

What is the purpose of this PR? is it an experiment?  can I close it?

Kevin

**Comment by dsyme** [comment]
@KevinRansom 

I initially wrote this to test the compiler perf scripts, and out of curiosity.

The technique omits castclass instructions wince we know from F# type safety that they are not going to fail. This technique seems to give a 1-2% perf gain in the compiler, but comes at the cost of generating unverifiable (though entirely valid, safe and correct - somewhat akin to the code emitted by Managed C++) .NET IL. 

1-2% gains are can add up.   Maybe we should embrace this technique. I'd like guidance about how you, @brettfo and @cartermp feel about it.  There would be some things to decide like whether we expose this as a user-facing code-generation feature.

**Comment by brettfo** [comment]
Is having unverifiable code going to cause problems when running on .NET Core?  Also, what about (the lack of) native code generation for F#?  F# already generates some IL patterns that aren't gracefully handled by some native code generators and this just broadens the gap, especially since the native code would be faster than the lack of `castclass` anyways.

Ultimately, I'm concerned that improving perf by 2% at the cost of potential runtime (and native code generation) isn't worth it, but I could be convinced otherwise.

**Comment by dsyme** [comment]
> Is having unverifiable code going to cause problems when running on .NET Core? 

No, I've no reason to believe that it would, unverifiable instructions are supported there

> Also, what about (the lack of) native code generation for F#? F# already generates some IL patterns that aren't gracefully handled by some native code generators and this just broadens the gap, especially since the native code would be faster than the lack of castclass anyways.

I don't think this effects that. Static verifiability of IL just isn't very important these days

> Ultimately, I'm concerned that improving perf by 2% at the cost of potential runtime (and native code generation) isn't worth it, but I could be convinced otherwise.

Unverifiable code is generated all the time by any tool doing native interop. 

Another way of looking at it is that emitting castclass instructions that are known to always succeed and are always nop's seems a bit of a waste of electricity in 2017 :) What are they for? To make us feel good that the code could in theory be statically verified to be typesafe, even if there's no purpose to that static verification? We would never emit these redundant checks for machine code for F# - why should we emit them in IL?  That puts IL at a permanent disadvantage over machine code

Basically no one in the .NET world even knows what static verifiabiity is for any more - it's just yet another relic from 1997 dreams, in this case of partial trust code execution (which certainly doesn't apply to the F# compiler, or really any .NET code these days)

**Comment by KevinRansom** [comment]
Coreclr runs unverifiable code fine.  However the code wouldn't work in partial trust environments, and AppDomains are coming back to coreclr in dotnetstandard  2.0.  Not sure if partial trust is coming back though.

I think emitting verifiable code by default is the best way forward, even at a performance cost.  Perhaps we can teach the jit to optimize it away.

**Comment by smoothdeveloper** [comment]
What about having the option (with a more precise name regarding `castclass`?) until the jit of the many CLI platforms can optimize it away?

Having that performance improvement for "free" based on that option sounds appealing.

**Comment by dsyme** [comment]
> I think emitting verifiable code by default is the best way forward. even at a performance cost. 

Emitting verifiable code from the compiler by default is correct.

However this is about an option to omit the needless costs of verifiable code, and using that option to emit the code of FSharp.Compiler.dll itself.  There is ultimately no solid rational argument why we aren't doing this, except 

1. The use of unverifiable code may mask some other code generation bug (However we could just verify the DLL compiled without the unverifiable code, or emit verifiable code in the debug build)

2. The performance gains aren't enough to bother. Even at 1-2% I'm tempted, every bit counts. However if we continue to make marginal improvements to the compiler elsewhere then we may see this rise to 3-5% of overall performance.  

> ...the code won't work in partial trust environments...

I've seen no information that partial trust is coming back to .NET. But in any case FSharp.Compiler.dll is never executed in partial trust mode (and wouldn't work if in partial trust mode anyway, it would need full permissions).

> Perhaps we can teach the JIT to optimize it away.

This will not be possible - the information is flowed across more complex boundaries than the JIT can reasonably analyze.  From what I know from the various JIT codebases, the JIT does not (and cannot - for performance and other reasons) analyze at the kind of type-based information needed to remove these instructions.

I'm not pushing for us to accept this immediately - we need to keep verifying the performance gain.  But ultimately we want F# to be competitive with native code, not paying unnecessary costs just to satisfy some legalistic bar of static verifiability of generated code. (As you can tell, I'm not a huge fan of the .NET verifiability rules for a modern high-performance IL - even though I helped craft them)

**Comment by KevinRansom** [comment]
@dsyme 

Have you reached any conclusions about this PR?  Do you want to continue iterating on it?

Kevin

**Comment by dsyme** [comment]
Let's leave it open.  The [perf results](https://github.com/Microsoft/visualfsharp/blob/master/tests/scripts/compiler-perf-results.txt) jump around all over the place still, which makes things hard, even though they are run on the same physical machine.  We need to come up with a measuring methodology (e.g. things like forcing GCs at and of particular phases) to measure things more accurately, this is a good PR to test any changes in that.

If we convincingly get a perf gain over 3% in type checking phase then let's accept it.

**Comment by dsyme** [comment]
Closing as we have collected enough perf data on this for now (the effect is not large but maybe worthwhile at some point)
